core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -49,7 +49,10 @@
             visit(map != null ? map.valueType() : null, schema.getValueType(), visitor));
 
       default:
-        return visitor.primitive(iType != null ? iType.asPrimitiveType() : null, schema);
+        if (iType == null || iType.isStructType()) {
+          return visitor.primitive(null, schema);
```

Comment:
When visitUnion, each element of UNION will be visited along with iType as StructType. Thus, we need to handle iType properly. Discussed offline, we decide to move the StructType handling logic to visitUnion so that no change need to be made here.